src/FSharp.Data.GraphQL.Server.AspNetCore/Serialization/JsonConverters.fs

```diff
@@ -0,0 +1,179 @@
+namespace FSharp.Data.GraphQL.Server.AspNetCore
+
+open FSharp.Data.GraphQL
+open System
+open System.Text.Json
+open System.Text.Json.Serialization
+
+[<Sealed>]
+type ClientMessageConverter<'Root>(executor : Executor<'Root>) =
+  inherit JsonConverter<ClientMessage>()
+
+  let raiseInvalidMsg explanation =
+    raise <| InvalidMessageException explanation
+
+  /// From the spec: "Receiving a message of a type or format which is not specified in this document will result in an immediate socket closure with the event 4400: &lt;error-message&gt;.
+  /// The &lt;error-message&gt; can be vaguely descriptive on why the received message is invalid."
+  let invalidMsg (explanation : string) =
+    InvalidMessage (4400, explanation)
+    |> Result.Error
+
+  let errMsgToStr (struct (docId: int, graphQLErrorMsgs: GQLProblemDetails list)) =
+   String.Join('\n', graphQLErrorMsgs |> Seq.map (fun err -> err.Message))
+
+  let unpackRopResult ropResult =
+    match ropResult with
+    | Ok x -> x
+    | Result.Error (InvalidMessage (_, explanation: string)) ->
+      raiseInvalidMsg explanation
+
+  let getOptionalString (reader : byref<Utf8JsonReader>) =
+    if reader.TokenType.Equals(JsonTokenType.Null) then
+      None
+    else
+      Some (reader.GetString())
+
+  let readPropertyValueAsAString (propertyName : string) (reader : byref<Utf8JsonReader>) =
+    if reader.Read() then
+      getOptionalString(&reader)
+    else
+      raiseInvalidMsg <| sprintf "was expecting a value for property \"%s\"" propertyName
+
+  let requireId (raw : RawMessage) : Result<string, ClientMessageProtocolFailure> =
+    match raw.Id with
+    | Some s -> Ok s
+    | None -> invalidMsg <| "property \"id\" is required for this message but was not present."
+
+  let requireSubscribePayload (serializerOptions : JsonSerializerOptions) (executor : Executor<'a>) (payload : JsonDocument option) : Result<GraphQLQuery, ClientMessageProtocolFailure> =
+    match payload with
+    | None ->
+      invalidMsg <| "payload is required for this message, but none was present."
+    | Some p ->
+      let rawSubsPayload = JsonSerializer.Deserialize<GraphQLRequest option>(p, serializerOptions)
+      match rawSubsPayload with
+      | None ->
+        invalidMsg <| "payload is required for this message, but none was present."
+      | Some subscribePayload ->
+        match subscribePayload.Query with
+        | None ->
+          invalidMsg <| sprintf "there was no query in the client's subscribe message!"
+        | Some query ->
+          query
+          |> GraphQLQueryDecoding.decodeGraphQLQuery serializerOptions executor subscribePayload.OperationName subscribePayload.Variables
+          |> Result.mapError (fun errMsg -> InvalidMessage (CustomWebSocketStatus.invalidMessage, errMsg |> errMsgToStr))
+
+
+  let readRawMessage (reader : byref<Utf8JsonReader>, options: JsonSerializerOptions) : RawMessage =
```

Comment:
Why not deserializing it to a DU as I did in my implementation? It will just be a declaration of DU without imperative stuff

Reply:
This is not a DU use case. Please note that all properties may (and often will) appear simultaneously in the JSON object.

Reply:
Ah, I think you mean the `connection_init` | `ping` | `pong` | ... part.
Ok, I have to admit that this was a first stab I took without thinking much about how it could be made different.
But, other than that, I suspect that my approach has a better performance 😄 .
Anyway, I will try to use the DU approach to see how it looks like.

Reply:
Ok, you handled that yourself, so I'm resolving this conversation.

Reply:
I decided to postpone that for later when I will add approach to perform integration tests on subscriptions